data_structures/stacks/prefix_evaluation.py

```diff
@@ -0,0 +1,60 @@
+"""
+Python3 program to evaluate a prefix expression.
+"""
+
+calc = {
+    "+": lambda x, y: x + y,
+    "-": lambda x, y: x - y,
+    "*": lambda x, y: x * y,
+    "/": lambda x, y: x / y,
+}
+
+
+def is_operand(c):
+    """
+    Return True if the given char c is an operand, e.g. it is a number
+
+    >>> is_operand("1")
+    True
+    >>> is_operand("+")
+    False
+    """
+    return c.isdigit()
+
+
+def evaluate(expression):
+    """
+    Evaluate a given expression in prefix notation.
+    Asserts that the given expression is valid.
+
+    >>> evaluate("+ 9 * 2 6")
+    21
+    >>> evaluate("/ * 10 2 + 4 1 ")
+    4.0
+    """
+    stack = []
+
+    # iterate over the string in reverse order
+    for c in expression.split()[::-1]:
+
+        # push operand to stack
+        if is_operand(c):
+            stack.append(int(c))
+
+        else:
+            # pop values from stack can calculate the result
+            # push the result onto the stack again
+            o1 = stack.pop()
+            o2 = stack.pop()
```

Comment:
What if there is only a single input in the string
`evaluate("9")`
A better error message would be useful.

Reply:
I think the current behaviour is mathematically correct. Just like in infix notation x = 9 is a valid expression. At least I think so. So `evaluate("9")` should output "9" as it does.

Reply:
Early morning sleep made me skip the entire logic :p

Reply:
☕️